test/Microsoft.AspNet.Razor.Test/TagHelpers/TagHelperParseTreeRewriterTest.cs

```diff
@@ -19,6 +19,866 @@ namespace Microsoft.AspNet.Razor.Test.TagHelpers
+                            new MarkupTagHelperBlock(
+                                "strong",
+                                attributes: new Dictionary<string, SyntaxTreeNode>
+                                {
+                                    ["custom"] = dateTimeNow
+                                },
+                                children: factory.Markup("words and spaces")))
+                    },
+                    {
+                        "<div class=\"btn\" />",
+                        new MarkupBlock(
+                            new MarkupTagBlock(
+                                factory.Markup("<div"),
+                                new MarkupBlock(
+                                    new AttributeBlockCodeGenerator(
+                                        name: "class",
+                                        prefix: new LocationTagged<string>(" class=\"", 4, 0, 4),
+                                        suffix: new LocationTagged<string>("\"", 15, 0, 15)),
+                                    factory.Markup(" class=\"").With(SpanCodeGenerator.Null),
+                                    factory.Markup("btn").With(
+                                        new LiteralAttributeCodeGenerator(
+                                            prefix: new LocationTagged<string>(string.Empty, 12, 0, 12),
+                                            value: new LocationTagged<string>("btn", 12, 0, 12))),
+                                    factory.Markup("\"").With(SpanCodeGenerator.Null)),
+                                factory.Markup(" />")))
+                    },
+                    {
+                        "<div class=\"btn\"></div>",
+                        new MarkupBlock(
+                            new MarkupTagBlock(
+                                factory.Markup("<div"),
+                                new MarkupBlock(
+                                    new AttributeBlockCodeGenerator(
+                                        name: "class",
+                                        prefix: new LocationTagged<string>(" class=\"", 4, 0, 4),
+                                        suffix: new LocationTagged<string>("\"", 15, 0, 15)),
+                                    factory.Markup(" class=\"").With(SpanCodeGenerator.Null),
+                                    factory.Markup("btn").With(
+                                        new LiteralAttributeCodeGenerator(
+                                            prefix: new LocationTagged<string>(string.Empty, 12, 0, 12),
+                                            value: new LocationTagged<string>("btn", 12, 0, 12))),
+                                    factory.Markup("\"").With(SpanCodeGenerator.Null)),
+                                factory.Markup(">")),
+                            blockFactory.MarkupTagBlock("</div>"))
+                    },
+                    {
+                        "<p random1=\"a\" class=\"btn\" />",
```

Comment:
please rename the `random1` attribute to something more meaningful like `notRequired` or `htmlAttribute`